projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -1,76 +1,75 @@
-import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+import { RenderableTopology, RenderableTopologyNode, TopologyEdge, TopologyNode } from "../../topology";
 
 export class GraphLayout {
 
-  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>([[0, []]]);
   private nodeToLevelMap = new Map<RenderableTopologyNode, number>();
-  private nodeCoOrdinatesMap: Map<TopologyNode, TopologyCoordinates> = new Map<TopologyNode, TopologyCoordinates>();
 
   public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
 
     this.findRootNodes(topology);
     this.fillNodeAndLevelMaps();
-    this.fillNodeCoOrdinatesMap();
-    this.setNodeCoOrdinates(topology);
+    this.assignCoordinatesToNodes();
   }
 
   private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>) {
     topology.nodes.forEach(node => {
       if (node.incoming.length === 0) {
         this.nodeToLevelMap.set(node, 0);
-        this.levelToNodesMap.has(0) ? this.levelToNodesMap.get(0)?.push(node) : this.levelToNodesMap.set(0, [node])
+        this.levelToNodesMap.get(0)?.push(node);
       }
     });
   }
 
   private fillNodeAndLevelMaps() {
 
-    const nodes = [...(this.levelToNodesMap.get(0) ?? [])];
-    const visited = new Set<TopologyNode>([...nodes.map(node => node.userNode)]);
+    const nodesThatAreGoingToBeExplored = [...(this.levelToNodesMap.get(0) ?? [])];
+    const nodesThatHaveBeenOrAreGoingToBeExplored = new Set<TopologyNode>([...nodesThatAreGoingToBeExplored.map(node => node.userNode)]);
 
-    while (nodes.length > 0) {
-      const parent = nodes[0];
-      const level = this.nodeToLevelMap.get(parent)! + 1;
+    this.levelOrderTraversal(nodesThatAreGoingToBeExplored, nodesThatHaveBeenOrAreGoingToBeExplored);
+  }
 
-      parent.outgoing.forEach(edge => {
-        const child = edge.target;
-        if (!visited.has(child.userNode)) {
-          visited.add(child.userNode);
-          nodes.push(child);
-          this.nodeToLevelMap.set(child, level);
-          this.levelToNodesMap.has(level) ? this.levelToNodesMap.get(level)?.push(child) : this.levelToNodesMap.set(level, [child]);
-        }
-      });
+  private levelOrderTraversal(nodes: RenderableTopologyNode[], visited: Set<TopologyNode>): void{
+    if(nodes.length === 0) {
+      return;
+    }
 
-      nodes.shift();
+    const parent = nodes[0];
+    nodes.shift();
```

Comment:
```suggestion
    const parent = node.shift();
```

(also, given that you're working with a graph not a tree, suggest renaming "parent" to "currentNode")

Reply:
actually, given that it's a graph, how do cycles look? Does it draw edges back across the nodes?